src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp

```diff
@@ -352,10 +354,11 @@ REDHAWK_PALEXPORT CONTEXT* PalAllocateCompleteOSContext(_Out_ uint8_t** contextB
     }
 
 #ifdef TARGET_X86
-    if (pfnRtlRestoreContext == NULL)
+    if (pfnRtlRestoreContext == NULL && pfnNtContinue == NULL)
     {
         HMODULE hm = GetModuleHandleW(_T("ntdll.dll"));
         pfnRtlRestoreContext = (PRTLRESTORECONTEXT)GetProcAddress(hm, "RtlRestoreContext");
+        pfnNtContinue = (PNTCONTINUE)GetProcAddress(hm, "NtContinue");
```

Comment:
It is not acceptable for .NET runtime to use undocumented Windows APIs. All Windows APIs used by .NET runtime must be exposed via Windows SDK and/or documented at https://learning.microsoft.com.

Reply:
That's rather unfortunate. I wasn't sure whether this still applies when there is an official API that superseded the undocumented one and which is available on up-to-date OSes.

SEH it is then. :-/

Reply:
I wonder if the RtlRestoreContext implementation for x86 from the coreclr PAL wouldn't be a better alternative then.

Reply:
I didn't see that one. It looks like a reasonable alternative.

Reply:
Do we have any pre-existing concept for sharing the PAL [assembly] code between CoreCLR and NativeAOT, or is this small enough to just copy the assembly?

Reply:
This is small enough and there is also Windows vs. Unix difference (we do not share asm code between Windows and Unix).

Reply:
> Do we have any pre-existing concept for sharing the PAL [assembly] code between CoreCLR and NativeAOT, or is this small enough to just copy the assembly?

AFAIK, nope https://github.com/dotnet/runtime/issues/64253. Would be neat if we start doing that by moving stuff to src/coreclr/minipal. :)

Reply:
Naïve port of the PAL code fails to pass the smoke tests: https://github.com/dotnet/runtime/commit/d08842ccb1f5e6f9d2f8a52014e42c4322b8127e

(I'll get back to it after sleeping on it)

Reply:
> Naïve port of the PAL code fails to pass the smoke tests: https://github.com/dotnet/runtime/commit/d08842ccb1f5e6f9d2f8a52014e42c4322b8127e

The PAL restore context is used for limited scenarios and it does not restore the whole context. It is not used to restart the thread at arbitrary point.

The implementation for Windows is expected to restore the whole context. It is used to restart the thread at arbitrary point. I think you would need to deal with restoring all AVX registers, potentially also restoring the x87 FPU stack, etc. A lot of code that is easy to make a subtle mistake in.

Reply:
I was suspecting that would be the case. Looks like I am stuck with SEH in the end.